Invalidate Session When Admin Resets Journalist Password

lev-csouffrant opened the pull request:
## Status

Ready for review

## Description of Changes

Fixes Issue #2300 - Failure to invalidate session when user's password is reset by admin.

Changes proposed in this pull request:

## Testing

This PR adds a new field to the Journalist model to include a "session nonce". This value is authenticated against the database when actions are made, and if it does not match the most recent value, the user is logged out.

I am also not 100% if I correctly wrote my database migration code, so I would feel more comfortable if someone made sure that was correct before pulling this in.

To test this:

1. Log in as a journalist
2. In a seperate browser, login as an admin. Change the password for the journalist.
3. Attempt to perform an action as the journalist in the first browser. It should flash that you have been logged out due to password change.

## Deployment

This commit modifies the models, and alembic migration code is written.

## Checklist

### If you made changes to the server application code:

- [X] Linting (`make lint`) and tests (`make -C securedrop test`) pass in the development container

### If you made non-trivial code changes:

- [X] I have written a test plan and validated it for this PR

redshiftzero commented:
thanks @lev-csouffrant! heads up that the lint failure here is due to our CI and not an issue you introduced, a fix is incoming in #4652

eloquence commented:
Hi @lev-csouffrant, the lint failure should be fixed in latest `develop`. Could you rebase your PR?

eloquence commented:
Thanks again for this PR @lev-csouffrant. We're currently very focused on landing v3 Onion Service support and preparing for the 1.0.0 release, so it may still take us a couple of weeks before we have cycles to review and test this change, but we'll get to it as soon as we can. In the meantime, please do rebase if you have a chance :)

redshiftzero commented on securedrop/tests/test_journalist.py:
```diff
@@ -427,13 +427,14 @@ def test_admin_cannot_delete_self(journalist_app, test_admin, test_journo):
 
 
 def test_admin_edits_user_password_success_response(journalist_app,
```
one thing we'd want for merge is test coverage for the case in this PR, basically: 

1. Test admin changes test user password
2. Test user visits page, gets 200 OK and "You have been logged out due to password change" in page text
3. User is indeed actually logged out, i.e. they are redirected to login upon trying to access another page, e.g. `assert "Login to access" in text`

redshiftzero commented on securedrop/tests/migrations/migration_60f41bb14d98.py:
```diff
@@ -0,0 +1,155 @@
+                    # be gone.
+                    assert journalist['session_nonce'] is None
+                except NoSuchColumnError:
+                    pass
```
migration test looks good!

redshiftzero commented on securedrop/alembic/versions/60f41bb14d98_add_session_nonce_to_journalist.py:
```diff
@@ -0,0 +1,78 @@
+                       id=:id""").bindparams(id=journalist.id)
+            )
+
+    # Now create new table with unique constraint applied.
```
nit: `# Now create new table with unique constraint applied.` -> `# Now create new table with null constraint applied.`

redshiftzero commented on securedrop/alembic/versions/60f41bb14d98_add_session_nonce_to_journalist.py:
```diff
@@ -0,0 +1,78 @@
+    # Add nonce column.
+    op.add_column('journalists_tmp', sa.Column('session_nonce', sa.Integer()))
+
+    # Add UUIDs to journalists_tmp table.
```
nit: `# Add UUIDs to journalists_tmp table.` -> `# Populate nonce column.`

lev-csouffrant commented:
Hmmm... I messed up this rebase somehow. :( Does someone think they can give me some advice in gitter PM on how to make this not horrifying? I'm not sure what to do to fix it. 

I fixed the couple issues pointed out, rebased, and added the test case requested. 

Also I'm unsure as to why that GUI test is failing. Doesn't seem like something I would have touched in my commit?
